packages/modeling/src/geometries/geom2/fromPoints.js

```diff
@@ -20,16 +20,9 @@ export const fromPoints = (points) => {
     throw new Error('the given points must define a closed geometry with three or more points')
   }
   // adjust length if the given points are closed by the same point
-  if (vec2.equals(points[0], points[length - 1])) --length
+  if (vec2.equals(points[0], points[length - 1])) points.length--
 
-  const sides = []
-  let prevpoint = points[length - 1]
-  for (let i = 0; i < length; i++) {
-    const point = points[i]
-    sides.push([vec2.clone(prevpoint), vec2.clone(point)])
```

Comment:
hmm... i'm wondering if this function should clone the points. i saw the cloning of vectors in project() below, so wanted to point this out for discussion.

Reply:
My opinion is that the correct action is never copy/clone unless necessary. That is kind of the core philosophy of an immutable data structure library. If we clone on every create call, we will never get benefits of shared vertex references, and there will be a huge number of unnecessary allocations.

Reply:
I would agree with that. With immutability we must copy on write, but no need to copy defensively. So copy is made of parts that would otherwise mutate.

Reply:
The usage is slightly different, as the points provided are those of the user, i.e. not part of a geometry. The create() function definitely expects the list of outlines to be immutable, which implies that the points are also immutable. But this is not the case for fromPoints(), where the provided points are outside the domain of the geometry.

By the way, we have changed some of the internal logic to create points, and use geom.create() instead of geom.fromPoints(). There's a significant performance boost due to those changes.

Also, the geometries should be consistent. The other geometries are cloning vectors.

Reply:
There are many places internal to JSCAD which also use `geom2.fromPoints`.

Also, I don't see documentation anywhere telling the user that `create` can use its input as part of its internal structure, but that `fromPoints` will clone everything. Both are available to users.

In fact, for `geom3`, both `create` and `fromPoints` will reuse user provided data without cloning. So I think the consistent behavior would be to assume immutability as I've done here.

I feel pretty strongly that we should assume the user will NOT modify the inputs they pass to JSCAD. We should document that assumption, and if a user really wants to modify things, they should clone them before passing to JSCAD. This gives us opportunity for maximum performance. And I think it is reasonable for users to assume if they modify things, it might affect the geometry.

I'll change it if you want in order to get the PR merged, but I really think it's better as-is.

Reply:
@platypii let's be consistent as least in functionality/design/documentation for the geometries.

- create(contents) expects the contents to be immutable (DONE)
- from*(data) functions should convert the data to into contents (the data is reused if possible and should not be modified), and then call create
- get rid of from*(data) functions that are just duplicating create()

IF another PR is required to correct the geometries then please work on that next as consistent geometry data structures / functions will improve V3 in many ways.

P.S. slice() needs to become a geometry as well, moving this into geometries.

Reply:
I have a subsequent PR ready which removes all uses of `fromPoints` which I'll put up next

Reply:
I think it should be removed entirely in the next PR because it is redundant with create.